### ISSUE TITLE
Audio drops out intermittently

### ISSUE THREAD

**Issue by andrewrk** [LMMS/lmms#158]
I get glitchy audio sound about half of the times I open LMMS. Usually it goes away and everything sounds normal again after tinkering around for 60 seconds or so. I am not yet able to reliably duplicate the problem or make it stop when it occurs.

**Comment by Umcaruje** [comment]
Is this still an issue? I've been using lmms with alsa for over 2 years now and I never encountered this. Can someone still reproduce?

**Comment by tresf** [comment]
Yes, this is still relevant for me and I can confirm that Apple has these identically described issues.  60 seconds is a bit extreme though.  Pausing and playing again after about 5 second pause fixes it every time.

**Comment by iraklisg** [comment]
I have the problem with "glitchy" sounds from time to time when I run LMMS with ALSA in linux mint 17 cinnamon 64bit. The "quick" workaround I have found is to reboot my pc.
Additionally I cannot figure out how to use LMMS with pulseaudio (the pulseaudio package is installed in my system but I cannot find "pulse" in Settings-> Audio interfaces...)

**Comment by CtrlFreak1337** [comment]
This issue only exists for me when I have additional programs open; but the problem is that I want to livestream my work. I need Firefox open to view stream chat, and OBS to actually stream it, but every time I make a major change to the project, the sound goes kaput. When I open Sound Settings to check, the alsa-plugin for LMMS is rapidly appearing and disappearing from the Applications tab.

**Comment by andrewrk** [comment]
@CtrlFreak1337 in addition to PulseAudio you can try using JACK to livestream your work.

**Comment by CtrlFreak1337** [comment]
@andrewrk Thanks. I've recently reinstalled LMMS, and the issue doesn't seem to be resurfacing. If it does, though, I'll be sure to look at those alternatives.

**Comment by trekkers2208** [comment]
I'm having this same issue.  Using Mint 17 Cinnamon, LMMS will work fine for a while, then all of a sudden while previewing instruments, the sound goes distorted.  I suspected having the browser open at the same time as LMMS caused the issue so i quit doing that.  Even when LMMS is the only program open it still does it.  Audio settings are set to use the OS card and my speakers selected as the device.

**Comment by jasp00** [comment]
I am unable to get any sound from PulseAudio. Is this still an issue with PulseAudio 8.0?

**Comment by jasp00** [comment]
Yes, it is. The `PULSE_LATENCY_MSEC=60` workaround does not work. The glitch mode is triggered by high CPU usage and it may last indefinitely.
The solution would be:

1. Detect the likely interception (ALSA device set to default and the presence of `/usr/share/alsa/alsa.conf.d/pulse.conf`).
2. Switch to the PulseAudio back end.

Would this be OK?

**Comment by noahb01** [comment]
Old but I may still have this issue on 1.2.0. The sound is incredibly distorted and glitchy (lots of high freq. noise, almost like you took a reverb on master and bitcrushed it), but it always goes away if I adjust the audio volume on my system (arch linux, kde + pulseaudio, sdl as LMMS backend)

**Comment by noahb01** [comment]
Also when this issue occurs it affects all system sounds, not just LMMS.

**Comment by stevphie** [comment]
This very easy to trigger, pick one of the default samples/presets available (by hold and click) while playing your track

**Comment by SadaleNet** [comment]
Now that I come to think of it, I suspect that it could be caused by CPU issue. If the CPU is at 100%, this may happen.

Anyway I'm not going to verify this theory any time soon. I haven't used LMMS for quite a while.

**Comment by dan-giddins** [comment]
> This very easy to trigger, pick one of the default samples/presets available (by hold and click) while playing your track

I can't recreate this issue on 1.3.0-alpha on windows. What's being described sounds a lot like issues with the selected audio driver, not LMMS. I'm going close this for now, if thats okay with everyone as there has been no activity on this issue for over a year. @tresf is this still an issue for you on mac?

**Comment by tresf** [comment]
I don't use Linux for audio production, but I doubt this was fixed.  I believe it's an underrun problem.  I patched this to the best of my ability on Windows by setting the lmms.exe process priority from normal to high.  I haven't done a similar patch to the Mac or Linux versions, so the QThread priorities are the best we have currently.

This should be closed by someone who actually uses Linux for production, and preferably a developer that does so on a slower machine.

Alternately we can close and tell users to request a reopen as needed, but I don't think this bug has been addressed in newer versions so I would credit faster machines for the fix, not any code changes.

**Comment by tresf** [comment]
Furthermore the patch proposed by @jasp00 has a link to a Debian mailing list, but no cross-reference to the ALSA list which would have needed to implement his patch.

So at time of writing this, even our own internal patch seems as if may have been merged prematurely.  @jasp00 would know more, but searching for the "don't intercept" `PUSLE_AUDIO_HOOK_CONF` introduced by #2292 in Google yields exactly two results, both mailing list archives between @jasp00 and the Debian mailing list.

https://www.mail-archive.com/pkg-pulseaudio-devel@lists.alioth.debian.org/msg05363.html

Regardless, in my testing, PulseAudio isn't the issue, just makes it worse.  I have a detailed write-up about this in another bug report, happy to locate and cross-link if needed.

**Comment by dan-giddins** [comment]
@tresf if no one is able to reproduce a bug, does it exsit...
or rather is it even a bug with LMMS?

> I suspect this issue is the problem of ALSA compatibility layer for pulse audio. I am not sure tho.

combined with

> Now that I come to think of it, I suspect that it could be caused by CPU issue.

and the fact that you solved it by changing the process priority to high implies that this is an issue with how the OS is handling the audio workload. What can we actually do about this for Mac or Linux?

**Comment by dan-giddins** [comment]
> I have a detailed write-up about this in another bug report, happy to locate and cross-link if needed.
That might be a good idea, incase this issue appears again in the future. Github needs an 'On-Hold' state for issues that can be neither reproduced by active developers, nor confirmed to be fixed...
I think its worth testing again to see if this still and issue if you are one of the people in this thread who is having this problem!

**Comment by tresf** [comment]
I produce with Mac, but a very fast machine, so hard to guage, but what I had discovered when I last investigated this is there are some  potential instances where our UI theading is incorrectly set as the same priority as the mixer, so something as simple as dragging a Window can make the underrun start and never recover.

I'm no expert on realtime DSP scheduling, nor a C++ coder, but I'll find and link my detailed findings last I looked into this...

**Comment by tresf** [comment]
@dan-giddins here's the thread: https://github.com/LMMS/lmms/issues/1600
Here's a relevant patch for the very-specific-issue found in that thread: https://github.com/LMMS/lmms/issues/1932

#1600 is a very good discussion because -- like this thread -- has a lot of mixed results.

My testing was performed here: https://github.com/LMMS/lmms/issues/1600#issuecomment-69782663

> Github really needs an 'On-Hold' state for issues that can be neither reproduced by active developers, nor confirmed to be fixed...

I started a similar cleanup effort here: https://github.com/LMMS/lmms/issues/4877, but people keep making bug reports faster than we can clean them up.

Because this is a pretty major bug, I see no reason to even suggest closing this until someone's fired up an old machine and done some testing.  This task only takes a few minutes.

What's important to understand is fast hardware as well as faster UI draw routines can mask stuff.  As a Java developer, I can state with certainty that Linux often exposes bad code that was working just fine on Windows.  It doesn't mean the the bug doesn't exist, but rather it's less likely to surface due to platform differences.

**Comment by RiedleroD** [comment]
I'm using a pentium-based Acer Travelmate B117M (low-powered enough?) as my main driver. My setup is Arch Linux 5.11.7 with wayland/sway and LMMS 1.2.2. I was compiling LMMS from the main branch to increase CPU load while testing this. The problem only occurs with the pulseaudio driver, not with ALSA, SDL or sndio. I tested all of them at least twice to make sure nothing weird is going on, since SDL uses pulseaudio afaik. I can also confirm that LMMS is the only program I use that has weird audio output on pulseaudio.

It does go away after a bit of tinkering. The first time, it was after dragging in the sf2 player, the second time after loading a vst. It seems long loading times like to stop the issue? Not sure, since, so far, I only managed to stop it twice.
